src/annotator/anchoring/range.js

```diff
@@ -0,0 +1,494 @@
+   * withing the bounds parameter. It returns an updated range with all
+   * properties updated. NOTE: Method returns null if all nodes fall outside
+   * of the bounds.
+   *
+   * bounds - An Element to limit the range to.
+   *
+   * Returns updated self or null.
+   */
+  limit(bounds) {
+    const nodes = $.grep(
+      this.textNodes(),
+      node => node.parentNode === bounds || $.contains(bounds, node.parentNode)
+    );
+    if (!nodes.length) {
+      return null;
+    }
+
+    this.start = nodes[0];
+    this.end = nodes[nodes.length - 1];
+
+    const startParents = $(this.start).parents();
+
+    for (let parent of Array.from($(this.end).parents())) {
+      if (startParents.index(parent) !== -1) {
+        this.commonAncestor = parent;
+        break;
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Convert this range into an object consisting of two pairs of (xpath,
+   * character offset), which can be easily stored in a database.
+   *
+   * root -           The root Element relative to which XPaths should be calculated
+   * ignoreSelector - A selector String of elements to ignore. For example
+   *                  elements injected by the annotator.
+   *
+   * Returns an instance of SerializedRange.
+   */
+  serialize(root, ignoreSelector) {
+    const serialization = function (node, isEnd) {
+      let origParent;
+      if (ignoreSelector) {
+        origParent = $(node).parents(`:not(${ignoreSelector})`).eq(0);
+      } else {
+        origParent = $(node).parent();
+      }
+      const xpath = xpathFromNode(origParent, root)[0];
+      const textNodes = getTextNodes(origParent);
+      // Calculate real offset as the combined length of all the
+      // preceding textNode siblings. We include the length of the
+      // node if it's the end node.
+      const nodes = textNodes.slice(0, textNodes.index(node));
+      let offset = 0;
+      for (let n of Array.from(nodes)) {
+        offset += n.nodeValue.length;
+      }
+
+      if (isEnd) {
+        return [xpath, offset + node.nodeValue.length];
+      } else {
+        return [xpath, offset];
+      }
+    };
+
+    const start = serialization(this.start);
+    const end = serialization(this.end, true);
+
+    return new Range.SerializedRange({
+      // XPath strings
+      start: start[0],
+      end: end[0],
+      // Character offsets (integer)
+      startOffset: start[1],
+      endOffset: end[1],
+    });
+  }
+
+  /**
+   * Creates a concatenated String of the contents of all the text nodes
+   * within the range.
+   *
+   * Returns a String.
+   */
+  text() {
+    return Array.from(this.textNodes())
+      .map(node => node.nodeValue)
+      .join('');
+  }
+
+  /**
+   * Fetches only the text nodes within the range.
+   *
+   * Returns an Array of TextNode instances.
+   */
+  textNodes() {
+    const textNodes = getTextNodes($(this.commonAncestor));
+    const [start, end] = Array.from([
```

Comment:
The `Array.from` is redundant here, as is the only array literal given that it is immediately destructured.